--- expat-1.95.6/configure.in~	Sat May 17 23:49:50 2003
+++ expat-1.95.6/configure.in	Sat May 17 23:49:50 2003
@@ -23,11 +23,7 @@
 dnl NOTE: esyscmd() is a GNU M4 extension. Thus, we wrap it in an appropriate
 dnl test. I believe this test will work, but I don't have a place with non-
 dnl GNU M4 to test it right now.
-define([expat_version], ifdef([__gnu__],
-                              [esyscmd(conftools/get-version.sh lib/expat.h)],
-                              [1.95.x]))
-AC_INIT(expat, expat_version, expat-bugs@mail.libexpat.org)
-undefine([expat_version])
+AC_INIT(expat, 1.95.6, expat-bugs@mail.libexpat.org)
 
 AC_CONFIG_SRCDIR(Makefile.in)
 AC_CONFIG_AUX_DIR(conftools)
@@ -50,9 +50,6 @@
 
 AC_CONFIG_HEADER(expat_config.h)
 
-sinclude(conftools/libtool.m4)
-sinclude(conftools/ac_c_bigendian_cross.m4)
-
 AC_LIBTOOL_WIN32_DLL
 AC_PROG_LIBTOOL
 
@@ -77,9 +77,7 @@
 
 dnl Checks for typedefs, structures, and compiler characteristics.
 
-dnl Note: Avoid using AC_C_BIGENDIAN because it does not
-dnl work in a cross compile.
-AC_C_BIGENDIAN_CROSS
+AC_C_BIGENDIAN
 
 AC_C_CONST
 AC_TYPE_OFF_T
