diff -uNpr misc.orig/test/mxc_pmic_test/pmic_testapp_battery/pmic_battery_test.c misc/test/mxc_pmic_test/pmic_testapp_battery/pmic_battery_test.c
--- misc.orig/test/mxc_pmic_test/pmic_testapp_battery/pmic_battery_test.c	2007-01-12 15:14:47.000000000 -0700
+++ misc/test/mxc_pmic_test/pmic_testapp_battery/pmic_battery_test.c	2007-02-06 15:10:15.000000000 -0700
@@ -226,7 +226,6 @@ int VT_pmic_batt_test(int switch_fct)
 		break;
 	case 1:
 		tst_resm(TINFO, "Test enable eol function of PMIC Battery");
-		tset2.threshold = 0;
 		for (i = 0; i < CHGR_OVOLT_THRESH_MAX; i++) {
 			tset2.typical = i;
 
