diff --exclude CVS -uNr netperf-2.4.0/src/netlib.c netperf-2.4.0.modified/src/netlib.c
--- netperf-2.4.0/src/netlib.c	2005-04-26 06:34:30.000000000 +0800
+++ netperf-2.4.0.modified/src/netlib.c	2006-04-19 14:18:53.000000000 +0800
@@ -545,6 +545,15 @@
     conv_rec.bytes[7-i] = scratch;
   }
   
+#if __FLOAT_WORD_ORDER != __BYTE_ORDER
+{
+    /* Fixup mixed endian floating point machines */
+    unsigned int scratch = conv_rec.words[0];
+    conv_rec.words[0] = conv_rec.words[1];
+    conv_rec.words[1] = scratch;
+}
+#endif
+
   return(conv_rec.whole_thing);
   
 }
@@ -579,6 +588,14 @@
     conv_rec.bytes[7-i] = scratch;
   }
   
+#if __FLOAT_WORD_ORDER != __BYTE_ORDER
+{
+    /* Fixup mixed endian floating point machines */
+    unsigned int scratch = conv_rec.words[0];
+    conv_rec.words[0] = conv_rec.words[1];
+    conv_rec.words[1] = scratch;
+}
+#endif
   /* we know that in the message passing routines htonl will */
   /* be called on the 32 bit quantities. we need to set things up so */
   /* that when this happens, the proper order will go out on the */
